src/operator/nnpack/nnpack_fully_connected-inl.h

```diff
@@ -0,0 +1,104 @@
+/*!
+ * Copyright (c) 2016 by Contributors
+ * \file nnpack_fully_connected-inl.h
+ * \brief
+ * \author Wei Wu
+*/
+#ifndef MXNET_OPERATOR_NNPACK_NNPACK_FULLY_CONNECTED_INL_H_
+#define MXNET_OPERATOR_NNPACK_NNPACK_FULLY_CONNECTED_INL_H_
+
+#include <dmlc/logging.h>
+#include <dmlc/parameter.h>
+#include <mxnet/operator.h>
+#include <algorithm>
+#include <map>
+#include <vector>
+#include <string>
+#include <utility>
+#include "../fully_connected-inl.h"
+#include "nnpack.h"
+#include "nnpack_util.h"
+
+namespace mxnet {
+namespace op {
+
+template <typename xpu, typename DType>
+class NNPACKFullyConnectedOp : public FullyConnectedOp<xpu, DType> {
+ private:
+  FullyConnectedParam param_;
+
+ public:
+  explicit NNPACKFullyConnectedOp(FullyConnectedParam p)
+      : FullyConnectedOp<xpu, DType>(p) {
+    this->param_ = p;
+  }
+
+ public:
+  virtual void Forward(const OpContext &ctx, const std::vector<TBlob> &in_data,
+                       const std::vector<OpReqType> &req,
+                       const std::vector<TBlob> &out_data,
+                       const std::vector<TBlob> &aux_args) {
+    using namespace mshadow;
+    using namespace mshadow::expr;
+    if (req[fullc::kOut] == kNullOp) return;
+    CHECK_EQ(req[fullc::kOut], kWriteTo);
+    size_t expected = param_.no_bias ? 2 : 3;
+    CHECK_EQ(in_data.size(), expected);
+    CHECK_EQ(out_data.size(), 1);
+    const TShape& ishape = in_data[fullc::kData].shape_;
+    const TShape& oshape = out_data[fullc::kOut].shape_;
+    Stream<xpu> *s = ctx.get_stream<xpu>();
+    Tensor<xpu, 2, DType> data = in_data[fullc::kData].get_with_shape<xpu, 2, DType>(
+        Shape2(ishape[0], ishape.ProdShape(1, ishape.ndim())), s);
+    Tensor<xpu, 2, DType> wmat = in_data[fullc::kWeight].get<xpu, 2, DType>(s);
+    Tensor<xpu, 2, DType> out = out_data[fullc::kOut].get_with_shape<xpu, 2, DType>(
+        Shape2(oshape[0], oshape.ProdShape(1, oshape.ndim())), s);
+    const size_t batch_size = data.shape_[0];
+    const size_t input_c = data.shape_[1];
+    bool use_nnpack{false};
+    // nnp_fully_connected_inference will do optimization for batch-size = 1
+    if (batch_size == 1) {
+      use_nnpack = true;
+    }
+    // nnp_fully_connected_output will do optimization for batch-size > 1
+    // but just found FullyConnected in NNPACK result is wrong when batch_size != 2^n
+    // so here only using NNPACK when batch_size = 2^n.
+    if ((batch_size > 1) && (!(batch_size & (batch_size - 1)))) {
```

Comment:
@piiswrong this cannot be done in fc op(but conv op can be done), because we cannot get batch size CreateOp, only if we change the interface like in conv.